pkg/apis/messaging/v1alpha1/pipeline_lifecycle.go

```diff
@@ -0,0 +1,165 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	duckv1alpha1 "github.com/knative/eventing/pkg/apis/duck/v1alpha1"
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/pkg/apis"
+	pkgduckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var pCondSet = apis.NewLivingConditionSet(PipelineConditionReady, PipelineConditionChannelsReady, PipelineConditionSubscriptionsReady, PipelineConditionAddressable)
+
+const (
+	// PipelineConditionReady has status True when all subconditions below have been set to True.
+	PipelineConditionReady = apis.ConditionReady
+
+	// PipelineChannelsReady has status True when all the channels created as part of
+	// this pipeline are ready.
+	PipelineConditionChannelsReady apis.ConditionType = "ChannelsReady"
+
+	// PipelineSubscriptionsReady has status True when all the subscriptions created as part of
+	// this pipeline are ready.
+	PipelineConditionSubscriptionsReady apis.ConditionType = "SubscriptionsReady"
+
+	// PipelineConditionAddressable has status true when this Pipeline meets
+	// the Addressable contract and has a non-empty hostname.
+	PipelineConditionAddressable apis.ConditionType = "Addressable"
+)
+
+// GetCondition returns the condition currently associated with the given type, or nil.
+func (ps *PipelineStatus) GetCondition(t apis.ConditionType) *apis.Condition {
+	return pCondSet.Manage(ps).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (ps *PipelineStatus) IsReady() bool {
+	return pCondSet.Manage(ps).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (ps *PipelineStatus) InitializeConditions() {
+	pCondSet.Manage(ps).InitializeConditions()
+}
+
+// PropagateSubscriptionStatuses sets the SubscriptionStatuses and PipelineConditionSubscriptionsReady based on
+// the status of the incoming subscriptions.
+func (ps *PipelineStatus) PropagateSubscriptionStatuses(subscriptions []*eventingv1alpha1.Subscription) {
+	ps.SubscriptionStatuses = make([]PipelineSubscriptionStatus, len(subscriptions))
+	allReady := true
+	// If there are no subscriptions, treat that as a False case. Could go either way, but this seems right.
+	if len(subscriptions) == 0 {
+		allReady = false
+
+	}
+	for i, s := range subscriptions {
+		ps.SubscriptionStatuses[i] = PipelineSubscriptionStatus{
+			Subscription: corev1.ObjectReference{
+				APIVersion: s.APIVersion,
+				Kind:       s.Kind,
+				Name:       s.Name,
```

Comment:
Namespace? (I'm not sure it is a great idea, but wanted to ask)

Reply:
done. here and below.